### PR TITLE
Implemented proof rerandomization

### PR DESCRIPTION
**Pull request by rozbb**
I need this functionality for a project I'm working on, and I figured it fits well into upstream. I also cleaned up the testing procedures.

### CODE REVIEW

**Comment by Pratyush** [comment]
Thanks for the PR! Do you have a link to the paper that proposes this randomization? It would be good to add this link to the comments.

File: src/test.rs [hunk 1]
```diff
@@ -112,6 +73,80 @@ mod cp6_782 {
...
+
+    // Check that the proofs are equivalent
+
+    assert!(verify_proof(&pvk, &proof1, &[c]).unwrap());
+    assert!(verify_proof(&pvk, &proof2, &[c]).unwrap());
+    assert!(verify_proof(&pvk, &proof3, &[c]).unwrap());
+
+    assert!(!verify_proof(&pvk, &proof1, &[E::Fr::zero()]).unwrap());
+    assert!(!verify_proof(&pvk, &proof2, &[E::Fr::zero()]).unwrap());
+    assert!(!verify_proof(&pvk, &proof3, &[E::Fr::zero()]).unwrap());
```
Comment (ValarDragon) [inline src/test.rs:120]: Can you add a check that none of the re-randomized proofs are the same proof?

File: src/prover.rs [hunk 2]
```diff
@@ -144,6 +144,32 @@ where
...
 
+/// Given a Groth16 proof, returns a proof of the same statement. The proof that is returned is
+/// computationally unlinkable to the given proof, assuming the RNG is cryptographically secure
+pub fn rerandomize_proof<E, R>(rng: &mut R, vk: &VerifyingKey<E>, proof: &Proof<E>) -> Proof<E>
+where
+    E: PairingEngine,
+    R: Rng,
+{
+    // These are our rerandomization factors
+    let (r, s) = (E::Fr::rand(rng), E::Fr::rand(rng));
```
Comment (ValarDragon) [inline src/prover.rs:155]: Note unrelated to this PR: We should add a method to `ark-std` for sampling non-zero elements

**Comment by ValarDragon** [comment]
Thanks for the new feature, and for making the tests test across multiple curves!

File: src/test.rs (hunk 1 above)
Comment (Pratyush) [reply src/test.rs:120]: This maybe even belongs in the rerandomize method, so that it panics if randomization fails.

Comment (rozbb) [reply src/test.rs:120]: I added it to the test. Why should this check be in the rerandomize method? The chance of an incomplete rerandomization is the chance that r_2 = 0, which is 2^(-255) or so

File: src/prover.rs [hunk 3]
```diff
@@ -144,24 +144,27 @@ where
     })
 }
 
-/// Given a Groth16 proof, returns a proof of the same statement. The proof that is returned is
-/// computationally unlinkable to the given proof, assuming the RNG is cryptographically secure
+/// Given a Groth16 proof, returns a fresh proof of the same statement. For a proof π of a
+/// statement S, the output of the non-deterministic procedure `rerandomize_proof(π)` is
+/// statistically indistinguishable from a fresh honest proof of S. For more info, see theorem 3 of
+/// BKSV20: https://eprint.iacr.org/2020/811
```
Comment (Pratyush) [inline src/prover.rs:150]:
```suggestion
/// [\[BKSV20\]](https://eprint.iacr.org/2020/811)
```

File: src/prover.rs [hunk 4]
```diff
@@ -144,24 +144,27 @@ where
...
 pub fn rerandomize_proof<E, R>(rng: &mut R, vk: &VerifyingKey<E>, proof: &Proof<E>) -> Proof<E>
 where
     E: PairingEngine,
     R: Rng,
 {
     // These are our rerandomization factors
-    let (r, s) = (E::Fr::rand(rng), E::Fr::rand(rng));
+    let (r1, r2) = (E::Fr::rand(rng), E::Fr::rand(rng));
```
Comment (Pratyush) [inline src/prover.rs:157]: Let's sample in a way that ensures that `r1` and `r2` can never be zero.

```suggestion
    let (mut r1, mut r2) = (E::Fr::zero(), E::Fr::zero());
    while r1.is_zero() || r2.is_zero() {
    	r1 = E::Fr::rand(rng);
    	r2 = E::Fr::rand(rng);
    }
```

Comment (rozbb) [reply src/prover.rs:157]: lgtm!

**Review by Pratyush — APPROVED** [review]

**Review by ValarDragon — APPROVED** [review]
LGTM! Thanks for upstreaming it

**Comment by burdges** [comment]
It's [noticeably faster](https://github.com/arkworks-rs/groth16/blob/master/src/prover.rs#L114) to create a non-zk proof and then rerandomize it afterwards.

**Pull request merged**